### ISSUE TITLE
Training on LRS3 or custom data

### ISSUE THREAD

**Issue by AvivSham** [ms-dot-k/Visual-Context-Attentional-GAN#3]
Hi All,
How are you?
Thank you for your wonderful and interesting work!
I'm trying to train VCA-GAN on LRS3 and wonder what steps I should follow for doing that. (what is the correct dir structure, etc.)

Cheers,

**Comment by ms-dot-k** [comment]
For training the network on LRS3, you should write your own dataset.py to load LRS3 dataset.
For the other code, you can use the model and training pipeline directly.

**Comment by AvivSham** [comment]
Thank you for your response. Is the preprocessing steps dataset agnostics? If some modifications are needed can you please name them?

Another question - what does the `speaker` argument stand for? in my case I want to train the model over all training set and not on a specific speaker, what happens if I don't pass this argument?

**Comment by ms-dot-k** [comment]
The uploaded preprocess step is for GRID.

For LRS3, the video is already cropped and contains the face. So you can skip the face alignment steps.
Instead, you can extract landmarks to crop the frames centered at the lips.
You can refer [here](https://github.com/facebookresearch/av_hubert/tree/main/avhubert/preparation)

The speaker argument is for selecting the speaker in GRID, it is only used when building file lists (build_file_list in src/data/vid_aud_grid.py) so you don't need it when writing the dataset.py for the LRS3 dataset.

In summary, I recommend you use different preprocessing steps for LRS3.
1. cropped the frames centered at the lips.
2. extract .wav from the video.
3. write your own dataset.py and train the model

**Comment by AvivSham** [comment]
thank you @ms-dot-k .
One last question, Are the video files loaded by `MultiDataset` in `vid_aud_grid.py` video files of the cropped lips?
If so what the dataset needs besides the cropped lips video is the corresponding wav files (in addition to some HPs of course).

**Comment by ms-dot-k** [comment]
In 'vid_aud_grid.py', the aligned face video is loaded and cropped with the bounding box of [59, 95, 195, 231] (defined in build_tensor function). However, if your video is already cropped, then you can remove this procedure and just use them directly. Also, your dataset should have lip-centered cropped video and corresponding wav files for training.

**Comment by AvivSham** [comment]
I will give it a try. In case it works I will create a pull request for you to merge. In the meantime please remain this issue open.

**Comment by AvivSham** [comment]
Hi @ms-dot-k ,
how are you?
it turns out that some of the videos in LRS3 are shorter than a sec (I checked and the shortest contains 12 frames). This fact raises an error about the window size which is set to 40 and is larger than the sequence. 
One way to address this issue would be to set the window size to 12 which differs from what was reported in the paper.
Does the current code support other solution? if not do you have another idea in mind?

**Comment by ms-dot-k** [comment]
Actually, I have the trained model on LRS3. But these days I am too busy to update it on the github.
However, I will try to upload the code and model. So can you wait for it?

**Comment by ms-dot-k** [comment]
I updated the codes for LRS3.
Let me know if you run into any issues.

**Comment by AvivSham** [comment]
Thank you for helping out. I will give it a try later today and let you know. 
I must say that it's possible to refactor the dataset classes to support all different datasets in a more elegant way.

**Comment by AvivSham** [comment]
Hi @ms-dot-k, can you please specify the librosa version you are using?

**Comment by ms-dot-k** [comment]
Hello, it is 0.8.0

**Comment by AvivSham** [comment]
I followed the LRS3 README and created the file structure as requested. The two main dirs `LRS3` and `LRS3_audio` are at the same level and inside each of them there are multiple dirs, one for each speaker. It seems like the dataset object tries to load the audio `wav` files from the wrong path containing the videos instead of the audio files. can you please check it out?

<img width="708" alt="image" src="https://user-images.githubusercontent.com/43371254/233829871-da557acd-6405-42b1-940e-71c33853647f.png">

**Comment by ms-dot-k** [comment]
The directory conversion from video to audio is hard coded in vid_aud_lrs3.py
https://github.com/ms-dot-k/Visual-Context-Attentional-GAN/blob/17c4a101bc16701038ad1b79dac6551109e4f2f1/src/data/vid_aud_lrs3.py#L134
It replaces 'LRS-TED' to 'LRS-TED_audio'. So if your directory name is LRS3, you should modify it.

**Comment by AvivSham** [comment]
Can you please share the approx runtime for training the model on LRS3? How many GPUs you used and what was the duration?

In addition, how many epochs you used for this dataset?

**Comment by AvivSham** [comment]
Also during val step received the following error: `Visual-Context-Attentional-GAN/src/data/vid_aud_lrs3.py", line 110, in build_tensor IndexError: index 160 is out of bounds for dimension 0 with size 160`
Can you look into it?

**Comment by ms-dot-k** [comment]
I trained 100 epochs by using a batch size of 32 with 2 A6000 GPUs are used. And it took about 1 week.
Also, I modified the dataset.py. The error is due to some samples exceeding the maximum length in the configuration.

**Comment by AvivSham** [comment]
@ms-dot-k thank you very much for your response.
I wonder if you can share the model weights? if you can't can you please send the HPs used to train the model?

**Comment by ms-dot-k** [comment]
Please find the last line in the readme_LRS.md

**Comment by AvivSham** [comment]
Thanks!
Final question, do you plan on releasing the code for "[Lip-to-Speech Synthesis in the Wild with Multi-task Learning](https://arxiv.org/abs/2302.08841)"?

**Comment by ms-dot-k** [comment]
Yes, we are preparing for it. It will take 3~6 weeks.

**Comment by AvivSham** [comment]
We would like to compare our approach with this method (the sub deadline is in 3 weeks) can you please send us weights and eval code?
we can continue this conversation via email: Mista2311@gmail.com

**Comment by AvivSham** [comment]
@ms-dot-k How are you?
did you see my previous message?

**Comment by ms-dot-k** [comment]
Yes, please check [here](https://github.com/ms-dot-k/Lip-to-Speech-Synthesis-in-the-Wild).
